### PR TITLE
Added a command to edit types using cmd.editor ##types

### LINKED ISSUE #13640: Add a command to edit types using cmd.editor
we need a command to ediit functoins, types, etc. in cmd.editor
something like tf write=-
and maybe a better way to define structs without having to quote the command all the time
i need a proposal for this, impl is easy, but cmd must be consistent
my idea is that.. now that we have proper C output for all the types loaded
we can dump that to a tmpfile
edit it
and reparse it again

### PR DESCRIPTION
**Pull request by sivaramaaa**
fix   #13640 

I have added this feature under command `toe`  aka "type open and edit" :D

UPDATE  :

ne

### CODE REVIEW

**Comment by sivaramaaa** [comment]
> how to edit a currently loaded type

U  mean to edit a specific type , then yeah a name as argument should be ok !

**Comment by radare** [comment]
looks red. pls fix

**Comment by sivaramaaa** [comment]
@radare i don't think  the red is coz of my PR , master is already RED

**Comment by radare** [comment]
Its green

File: libr/core/cmd_type.c [hunk 1]
```diff
@@ -1232,6 +1233,25 @@ static int cmd_type(void *data, const char *input) {
 					sdb_close (db_tmp);
 					sdb_free (db_tmp);
 				}
+			}  else if (input[1] == 'e') { // toe
+				char *str = r_core_cmd_str (core , "tc");
+				char *tmp = r_core_editor (core, "*.h", str);
```
Comment (radare) [inline libr/core/cmd_type.c]: Not sure if we want to edit all the includes

File: libr/core/cmd_type.c [hunk 2]
```diff
@@ -1232,6 +1233,25 @@ static int cmd_type(void *data, const char *input) {
...
+					char *out = r_parse_c_string (core->anal, tmp, &error_msg);
+					if (out) {
+						// remove previous types and save new edited types
+						sdb_reset (TDB);
+						r_parse_reset ();
+						r_anal_save_parsed_type (core->anal, out);
+						free (out);
+					}
+					if (error_msg) {
+						fprintf (stderr, "%s", error_msg);
```
Comment (radare) [inline libr/core/cmd_type.c]: Missing newline

File: libr/core/cmd_type.c [hunk 3]
```diff
@@ -1232,6 +1233,25 @@ static int cmd_type(void *data, const char *input) {
 					sdb_close (db_tmp);
 					sdb_free (db_tmp);
 				}
+			}  else if (input[1] == 'e') { // toe
+				char *str = r_core_cmd_str (core , "tc");
```
Comment (radare) [inline libr/core/cmd_type.c]: Str is never freed

File: libr/core/cmd_type.c (hunk 1 above)
Comment (sivaramaaa) [reply libr/core/cmd_type.c]: No , afaik the `*.h` means to create a temporary file with ".h" extension !

Comment (radare) [reply libr/core/cmd_type.c]: right :) i should pick a different char for that

File: libr/core/cmd_type.c [hunk 4]
```diff
@@ -1232,6 +1233,26 @@ static int cmd_type(void *data, const char *input) {
...
+					if (out) {
+						// remove previous types and save new edited types
+						sdb_reset (TDB);
+						r_parse_reset ();
+						r_anal_save_parsed_type (core->anal, out);
+						free (out);
+					}
+					if (error_msg) {
+						fprintf (stderr, "%s\n", error_msg);
+						free (error_msg);
```
Comment (radare) [inline libr/core/cmd_type.c]: why not use eprintf?

File: libr/core/cmd_type.c [hunk 5]
```diff
@@ -24,6 +24,7 @@ static const char *help_msg_t[] = {
 	"tn", "[?] [-][addr]", "manage noreturn function attributes and marks",
 	"to", " -", "Open cfg.editor to load types",
 	"to", " <path>", "Load types from C header file",
+	"toe", "", "Open cfg.editor to edit types",
```
Comment (radare) [inline libr/core/cmd_type.c]: isnt toe taking an argument to specify the name of the struct to edit?.. well, struct or enum or function, or whatever its named like this

File: libr/core/cmd_type.c (hunk 4 above)
Comment (sivaramaaa) [reply libr/core/cmd_type.c]: yeah ,  i will change it !

File: libr/core/cmd_type.c (hunk 5 above)
Comment (sivaramaaa) [reply libr/core/cmd_type.c]: Not yet , it's still not implemented :D , will do it today !

**Comment by radare** [comment]
ok cool! will look again after that!

**Comment by sivaramaaa** [comment]
@radare  i have implemented the feature , we can merge this after ur review , it's red as master is red , not related to this PR

**Pull request merged**